### PR TITLE
[JsonPatch]: Removed Test operation code and changed T to TModel

### PR DESCRIPTION
**Pull request by kirthik**
@Eilon, @rynowak - Could you please take a look?

### CODE REVIEW

File: src/Microsoft.AspNet.JsonPatch/Adapters/IObjectAdapter.cs [hunk 1]
```diff
@@ -8,14 +8,14 @@ namespace Microsoft.AspNet.JsonPatch.Adapters
...
-        void Move(Operation<T> operation, T objectToApplyTo);
-        void Remove(Operation<T> operation, T objectToApplyTo);
-        void Replace(Operation<T> operation, T objectToApplyTo);
-        void Test(Operation<T> operation, T objectToApplyTo);
+        void Add(Operation<TModel> operation, TModel objectToApplyTo);
+        void Copy(Operation<TModel> operation, TModel objectToApplyTo);
+        void Move(Operation<TModel> operation, TModel objectToApplyTo);
+        void Remove(Operation<TModel> operation, TModel objectToApplyTo);
+        void Replace(Operation<TModel> operation, TModel objectToApplyTo);
+        void Test(Operation<TModel> operation, TModel objectToApplyTo);
```
Comment (rynowak) [inline src/Microsoft.AspNet.JsonPatch/Adapters/IObjectAdapter.cs]: ???

Comment (kirthik) [inline src/Microsoft.AspNet.JsonPatch/Adapters/IObjectAdapter.cs]: I kept `Test` API but just removed the code in `ObjectAdapter` and added an error that says "test operation is not supported"

Comment (rynowak) [inline src/Microsoft.AspNet.JsonPatch/Adapters/IObjectAdapter.cs]: What kind of error will a user see? `ModelState` or 500/Exception?

Comment (kirthik) [inline src/Microsoft.AspNet.JsonPatch/Adapters/IObjectAdapter.cs]: If user calls `ApplyTo(patchDoc, ModelState)` then he will see a `ModelState` error. If he calls `ApplyTo(patchDoc)` then he will see `JsonPatchException`

Comment (yishaigalatzer) [inline src/Microsoft.AspNet.JsonPatch/Adapters/IObjectAdapter.cs]: Should probably see UnsupportedException

Comment (kirthik) [inline src/Microsoft.AspNet.JsonPatch/Adapters/IObjectAdapter.cs]: Sure. I will change `JsonPatchException` to `UnSupportedException`.

File: src/Microsoft.AspNet.JsonPatch/TypedJsonPatchDocument.cs [hunk 2]
```diff
@@ -64,12 +65,18 @@ public JsonPatchDocument<T> Add<TProp>(Expression<Func<T, TProp>> path, TProp va
...
+            IList<TProp>>> path,
+            TProp value,
+            int position)
         {
-            Operations.Add(new Operation<T>(
+            Operations.Add(new Operation<TModel>(
                 "add",
-                ExpressionHelpers.GetPath<T, IList<TProp>>(path).ToLower() + "/" + position,
-                null, value));
+                ExpressionHelpers.GetPath(path).ToLower() + "/" + position,
```
Comment (Eilon) [inline src/Microsoft.AspNet.JsonPatch/TypedJsonPatchDocument.cs]: `ToLower()` --> `ToLowerInvariant()`. Please check for all occurences.

Comment (Eilon) [inline src/Microsoft.AspNet.JsonPatch/TypedJsonPatchDocument.cs]: (Can do it in a separate PR if you want.)

**Comment by Eilon** [comment]
:shipit: from me. (Can do my comment in a separate PR so it's not too many changes in this PR.)

File: src/Microsoft.AspNet.JsonPatch/TypedJsonPatchDocument.cs (hunk 2 above)
Comment (kirthik) [inline src/Microsoft.AspNet.JsonPatch/TypedJsonPatchDocument.cs]: Yes, I will make this change in a separate PR.

File: src/Microsoft.AspNet.JsonPatch/Adapters/IObjectAdapter.cs [hunk 3]
```diff
@@ -8,14 +8,13 @@ namespace Microsoft.AspNet.JsonPatch.Adapters
     /// <summary>
     /// Defines the operations that can be performed on a JSON patch document.
     /// </summary>
-    /// <typeparam name="T"></typeparam>
-    public interface IObjectAdapter<T> where T : class
+    /// <typeparam name="TModel">The type of the model.</typeparam>
+    public interface IObjectAdapter<TModel> where TModel : class
```
Comment (rynowak) [inline src/Microsoft.AspNet.JsonPatch/Adapters/IObjectAdapter.cs]: This is an improvement, but we should still think about where in code the `TModel` really even makes sense. Just because you have a type doesn't mean you have to be generic (if that makes sense).

**Comment by rynowak** [comment]
Don't undo it now, but my advice in the future would be to send 2 PRs for this. The `T` -> `TModel` change is mechanical and could be reviewed by someone in 30 seconds.

File: src/Microsoft.AspNet.JsonPatch/Operations/TypedOperation.cs [hunk 4]
```diff
@@ -1,11 +1,12 @@
 ﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using Microsoft.AspNet.JsonPatch.Adapters;
 
 namespace Microsoft.AspNet.JsonPatch.Operations
 {
-    public class Operation<T> : Operation where T : class
+    public class Operation<TModel> : Operation where TModel : class
```
Comment (rynowak) [inline src/Microsoft.AspNet.JsonPatch/Operations/TypedOperation.cs]: Another note for cleanup - our guidelines say this file should be called `OperationOfT.cs`

**Comment by rynowak** [comment]
:shipit:

File: src/Microsoft.AspNet.JsonPatch/Adapters/IObjectAdapter.cs (hunk 3 above)
Comment (kirthik) [inline src/Microsoft.AspNet.JsonPatch/Adapters/IObjectAdapter.cs]: Sure. I will do this exercise in a separate PR.

File: src/Microsoft.AspNet.JsonPatch/Operations/TypedOperation.cs (hunk 4 above)
Comment (kirthik) [inline src/Microsoft.AspNet.JsonPatch/Operations/TypedOperation.cs]: Will change file names.

**Comment by kirthik** [comment]
Fix checked in - 3042a62ba95c014179dd5604dc10fecfbd4c924a
